lib/datadog/kit/identity.rb

```diff
@@ -0,0 +1,36 @@
+# typed: false
+
+module Datadog
+  module Kit
+    # Tracking identity via traces
+    module Identity
+      # Attach user information to the trace
+      #
+      # Values must be strings. Keys are free form, although the following keys have specific meanings:
+      # - id: Mandatory. Username or client id extracted from the access token or
+      #   Authorization header in the inbound request from outside the system.
+      # - email: Email of the authenticated user associated to the trace.
+      # - name: User-friendly name. To be displayed in the UI if set.
+      # - session_id: Session ID of the authenticated user.
+      # - role: Actual/assumed role the client is making the request under
+      #   extracted from token or application security context.
+      # - scope: Scopes or granted authorities the client currently possesses
+      #   extracted from token or application security context. The value would
+      #   come from the scope associated with an OAuth 2.0 Access Token or an
+      #   attribute value in a SAML 2.0 Assertion.
+      def self.set_user(trace, data = {})
```

Comment:
I was wondering if the argument `data = {}` was a good pattern and ran a few benchmarks:
```ruby
require 'benchmark/ips'

def x(arg = {}) end
def y(**arg) end
def z(
  id: nil,
  email: nil,
  name: nil,
  session_id: nil,
  role: nil,
  scope: nil
) end

puts "Passing no parameters"

Benchmark.ips do |x|
  x.report("arg = {}") { x }
  x.report("**arg") { y }
  x.report("a1:nil, a2:nil, ...") { z }

  x.compare!
end

puts "Passing all parameters"

Benchmark.ips do |x|
  x.report("arg = {}") { x(id: 1,
                           email: 1,
                           name: 1,
                           session_id: 1,
                           role: 1,
                           scope: 1) }
  x.report("**arg") { y(id: 1,
                        email: 1,
                        name: 1,
                        session_id: 1,
                        role: 1,
                        scope: 1) }
  x.report("a1:nil, a2:nil, ...") { z(
    id: 1,
    email: 1,
    name: 1,
    session_id: 1,
    role: 1,
    scope: 1) }

  x.compare!
end

# Passing no parameters
# Warming up --------------------------------------
#             arg = {}     2.612M i/100ms
#                **arg   821.787k i/100ms
#  a1:nil, a2:nil, ...     1.711M i/100ms
# Calculating -------------------------------------
#             arg = {}     23.538M (±16.3%) i/s -    114.943M in   5.082824s
#                **arg     10.327M (± 7.0%) i/s -     51.773M in   5.041600s
#  a1:nil, a2:nil, ...     17.167M (± 1.7%) i/s -     87.276M in   5.085654s
# 
# Comparison:
#             arg = {}: 23537793.6 i/s
#  a1:nil, a2:nil, ...: 17166705.3 i/s - 1.37x  (± 0.00) slower
#                **arg: 10327006.9 i/s - 2.28x  (± 0.00) slower
# 
# Passing all parameters
# Warming up --------------------------------------
#             arg = {}   402.644k i/100ms
#                **arg   337.917k i/100ms
#  a1:nil, a2:nil, ...     1.105M i/100ms
# Calculating -------------------------------------
#             arg = {}      4.453M (± 3.3%) i/s -     22.548M in   5.069373s
#                **arg      3.371M (± 1.7%) i/s -     16.896M in   5.013207s
#  a1:nil, a2:nil, ...     11.078M (± 2.2%) i/s -     56.332M in   5.087385s
# 
# Comparison:
#  a1:nil, a2:nil, ...: 11078320.1 i/s
#             arg = {}:  4453120.8 i/s - 2.49x  (± 0.00) slower
#                **arg:  3371254.7 i/s - 3.29x  (± 0.00) slower
```

I'd say we should use explicit keyword arguments, defaulting them to `nil` if optional, if possible.

Unless this method is overwhelmingly invoked with no keyword arguments (which does not seem to be the case for `set_user`).
Or if this method is commonly invoked with "extra" keyword arguments, something we've been trying to avoid in `ddtrace`.

Reply:
Thanks for the benchmarks!

Most of the cases should be one of the listed arguments.

> Unless this method is overwhelmingly invoked with no keyword arguments (which does not seem to be the case for set_user).

`id` is mandatory so that settles this one.

> Or if this method is commonly invoked with "extra" keyword arguments, something we've been trying to avoid in ddtrace.

The RFC for user attribution specifies that it should be able to be called with arbitrary additional keys, resulting in `usr.#{key}` tags.

Reply:
> The RFC for user attribution specifies that it should be able to be called with arbitrary additional keys

I'll check with the team whether that is truly useful and required, as I vastly prefer the explicit keyword argument version.

Also, I just thought it could be written with a mixed form, although I have to check if that behaves consistently across Ruby 2.1 to 3.1:

```
irb(main):001:0> def foo(x, a:, b: nil, **o); p x; p a; p b; p o; end
=> :foo
irb(main):002:0>  foo(1, c: 2, a: 3, b: 4)
1
3
4
{:c=>2}
=> {:c=>2}
```

Perf results (oddly when passing values it performs like explicit ones even though there are extra args):

```ruby
require 'benchmark/ips'

def w(id: nil, email: nil, **arg) end
def x(arg = {}) end
def y(**arg) end
def z(
  id: nil,
  email: nil,
  name: nil,
  session_id: nil,
  role: nil,
  scope: nil
) end

puts "Passing no parameters"

Benchmark.ips do |x|
  x.report("arg = {}") { x }
  x.report("**arg") { y }
  x.report("a1:nil, a2:nil, ...") { z }
  x.report("a1:, a2: nil, **arg") { w }

  x.compare!
end

puts "Passing all parameters"

Benchmark.ips do |x|
  x.report("arg = {}") { x(id: 1,
                           email: 1,
                           name: 1,
                           session_id: 1,
                           role: 1,
                           scope: 1) }
  x.report("**arg") { y(id: 1,
                        email: 1,
                        name: 1,
                        session_id: 1,
                        role: 1,
                        scope: 1) }
  x.report("a1:nil, a2:nil, ...") { z(
    id: 1,
    email: 1,
    name: 1,
    session_id: 1,
    role: 1,
    scope: 1) }
  x.report("a1:, a2:nil, **arg") { z(
    id: 1,
    email: 1,
    name: 1,
    session_id: 1,
    role: 1,
    scope: 1) }

  x.compare!
end

# Passing no parameters
# Warming up --------------------------------------
#             arg = {}     2.259M i/100ms
#                **arg   974.352k i/100ms
#  a1:nil, a2:nil, ...     1.396M i/100ms
#  a1:, a2: nil, **arg   858.774k i/100ms
# Calculating -------------------------------------
#             arg = {}     22.719M (± 1.6%) i/s -    115.205M in   5.072289s
#                **arg      9.758M (± 0.9%) i/s -     49.692M in   5.092984s
#  a1:nil, a2:nil, ...     13.936M (± 1.1%) i/s -     69.816M in   5.010329s
#  a1:, a2: nil, **arg      8.561M (± 1.2%) i/s -     42.939M in   5.016145s
# 
# Comparison:
#             arg = {}: 22718715.5 i/s
#  a1:nil, a2:nil, ...: 13936055.5 i/s - 1.63x  (± 0.00) slower
#                **arg:  9757828.5 i/s - 2.33x  (± 0.00) slower
#  a1:, a2: nil, **arg:  8561271.5 i/s - 2.65x  (± 0.00) slower
# 
# Passing all parameters
# Warming up --------------------------------------
#             arg = {}   386.356k i/100ms
#                **arg   373.197k i/100ms
#  a1:nil, a2:nil, ...   944.971k i/100ms
#   a1:, a2:nil, **arg   946.007k i/100ms
# Calculating -------------------------------------
#             arg = {}      3.879M (± 0.9%) i/s -     19.704M in   5.080762s
#                **arg      3.719M (± 1.6%) i/s -     18.660M in   5.019186s
#  a1:nil, a2:nil, ...      9.371M (± 4.3%) i/s -     47.249M in   5.055286s
#   a1:, a2:nil, **arg      9.456M (± 0.8%) i/s -     47.300M in   5.002735s
# 
# Comparison:
#   a1:, a2:nil, **arg:  9455565.6 i/s
#  a1:nil, a2:nil, ...:  9370838.4 i/s - same-ish: difference falls within error
#             arg = {}:  3878537.6 i/s - 2.44x  (± 0.00) slower
#                **arg:  3718797.3 i/s - 2.54x  (± 0.00) slower
# 
```

Reply:
Mixed form seems to work fine, although with explicit keyword args can't distinguish between an absent key and a `nil`-set key.